### PR TITLE
STRATEGY: Detect abnormal volume increases using isolation forest

### PR DESCRIPTION
**Pull request by narumiruna**
This pull request introduces a new trading strategy named "Sentinel" and includes various changes to support its implementation. The most important changes include the addition of the configuration for the Sentinel strategy, updates to dependencies, and the implementation of the strategy itself.

### Configuration and Dependencies:

* [`config/sentinel.yaml`](diffhunk://#diff-567a2873282a19afd4453422fa4ba2e0dcd8ff8a185d784e9570f5fb6e01f3f4R1-R19): Added configuration for sessions, persistence, and exchange strategies, including settings for the Sentinel strategy.
* [`go.mod`](diffhunk://#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R128): Added `github.com/narumiruna/go-iforest` as an indirect dependency.
* [`pkg/cmd/strategy/builtin.go`](diffhunk://#diff-4b4599b9a4be6b33da6400a2f3ea0d0942210bb05a3cd23e1978fc52e42f5456R43): Imported the Sentinel strategy to make it available for use.

### Sentinel Strategy Implementation:

* [`pkg/strategy/sentinel/strategy.go`](diffhunk://#diff-fb45e392467bc3eb90a946d3959f366997f74ccea64ed72fb9d75077c03865efR1-R161): Implemented the Sentinel strategy, including methods for initialization, default settings, subscription to market data, and running the strategy. The strategy uses an Isolation Forest algorithm to detect anomalies in trading volumes.

### Utility Functions:

* [`pkg/datatype/floats/slice.go`](diffhunk://#diff-a141ba682a66a847586b54b354ce469cd573f1036d4f397f86a9eeed150f225aR90-R107): Added `Var` and `Std` methods to the `Slice` type to calculate variance and standard deviation, which are used in the Sentinel strategy.

### CODE REVIEW

**Review by c9s — APPROVED** [review]

File: pkg/strategy/sentinel/strategy.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+func (s *Strategy) notifyIsolationForestScore(score float64) {
+	now := time.Now()
+	if now.Sub(s.lastNotifyTime) < s.NotificationInterval {
+		log.Infof("Skipping notification for symbol: %s, score: %f, due to short interval", s.Symbol, score)
+		return
+	}
+	s.lastNotifyTime = now
+
+	if channel, ok := bbgo.Notification.RouteSymbol(s.Symbol); ok {
```
Comment (c9s) [inline pkg/strategy/sentinel/strategy.go:153]: RouteSymbol will be deprecated

File: pkg/strategy/sentinel/strategy.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+	lastScore := scores[len(scores)-1]
+	log.Warnf("Symbol: %s, isolation forest score: %f, threshold: %f, kline: %s", s.Symbol, lastScore, s.ScoreThreshold, kline.String())
+	if lastScore > s.ScoreThreshold {
+		s.notifyIsolationForestScore(lastScore)
+	}
+}
+
+func (s *Strategy) notifyIsolationForestScore(score float64) {
+	now := time.Now()
+	if now.Sub(s.lastNotifyTime) < s.NotificationInterval {
```
Comment (c9s) [inline pkg/strategy/sentinel/strategy.go:147]: you can use rate limiter?

File: pkg/strategy/sentinel/math.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package sentinel
+
+import "math"
+
+func mean(floats []float64) float64 {
```
Comment (c9s) [inline pkg/strategy/sentinel/math.go:5]: I think we have implemented the "mean" somewhere for float64 slice, you can grep

File: pkg/strategy/sentinel/math.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+	var sum float64
+	for _, f := range floats {
+		sum += f
+	}
+	return sum / float64(len(floats))
+}
+
+func calculateMovingMeanAndStdDev(samples []float64, length int) (float64, float64) {
```
Comment (c9s) [inline pkg/strategy/sentinel/math.go:13]: we also have the std dev?

File: pkg/strategy/sentinel/strategy.go (hunk 1 above)
Comment (narumiruna) [reply pkg/strategy/sentinel/strategy.go:153]: ok

File: pkg/strategy/sentinel/math.go (hunk 4 above)
Comment (narumiruna) [reply pkg/strategy/sentinel/math.go]: I will add std dev method for floats.Slice

File: pkg/strategy/sentinel/math.go (hunk 3 above)
Comment (narumiruna) [reply pkg/strategy/sentinel/math.go]: replaced by floats.Slice

**Review by c9s — APPROVED** [review]

**Pull request merged**